tedana/metrics/collect.py

```diff
@@ -0,0 +1,140 @@
+    """
+    Fit TE-dependence and -independence models to components.
+
+    Parameters
+    ----------
+    data_cat : (S x E x T) array_like
+        Input data, where `S` is samples, `E` is echos, and `T` is time
+    data_optcom : (S x T) array_like
+        Optimally combined data
+    mixing : (T x C) array_like
+        Mixing matrix for converting input data to component space, where `C`
+        is components and `T` is the same as in `data_cat`
+    mask : img_like
+        Mask
+    tes : list
+        List of echo times associated with `data_cat`, in milliseconds
+    ref_img : str or img_like
+        Reference image to dictate how outputs are saved to disk
+    reindex : bool, optional
+        Whether to sort components in descending order by Kappa. Default: False
+    mixing_z : (T x C) array_like, optional
+        Z-scored mixing matrix. Default: None
+    algorithm : {'kundu_v2', 'kundu_v3', None}, optional
+        Decision tree to be applied to metrics. Determines which maps will be
+        generated and stored in seldict. Default: None
+    label : :obj:`str` or None, optional
+        Prefix to apply to generated files. Default is None.
+    out_dir : :obj:`str`, optional
+        Output directory for generated files. Default is current working
+        directory.
+    verbose : :obj:`bool`, optional
+        Whether or not to generate additional files. Default is False.
+
+    Returns
+    -------
+    comptable : (C x X) :obj:`pandas.DataFrame`
+        Component metric table. One row for each component, with a column for
+        each metric. The index is the component number.
+    seldict : :obj:`dict` or None
+        Dictionary containing component-specific metric maps to be used for
+        component selection. If `algorithm` is None, then seldict will be None as
+        well.
+    betas : :obj:`numpy.ndarray`
+    mmix_new : :obj:`numpy.ndarray`
+    """
+    RepLGR.info('The following metrics were calculated: {}.'.format(', '.join(metrics)))
+
+    if not (data_cat.shape[0] == data_optcom.shape[0] == mask.sum()):
+        raise ValueError('First dimensions (number of samples) of data_cat ({0}), '
+                         'data_optcom ({1}), and mask ({2}) do not '
+                         'match'.format(data_cat.shape[0], data_optcom.shape[0],
+                                        mask.shape[0]))
+    elif data_cat.shape[1] != len(tes):
+        raise ValueError('Second dimension of data_cat ({0}) does not match '
+                         'number of echoes provided (tes; '
+                         '{1})'.format(data_cat.shape[1], len(tes)))
+    elif not (data_cat.shape[2] == data_optcom.shape[1] == mixing.shape[0]):
+        raise ValueError('Number of volumes in data_cat ({0}), '
+                         'data_optcom ({1}), and mixing ({2}) do not '
+                         'match.'.format(data_cat.shape[2], data_optcom.shape[1], mixing.shape[0]))
+
+    mixing = mixing.copy()
```

Comment:
I think this copy is unnecessary, here, since the next time it's used is providing `mixing` to `flip_components ()` which will trigger a copy.

Reply:
But we *should* add a check for `if mixing_z is None` and assign `mixing` to `mixing_z`.